UPX compressed amd64 executable crashes on macOS Ventura 13.0

This issue tracker is **ONLY** used for reporting bugs.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?

The UPX compressed amd64 executable crashes on macOS 13.0.
I made a small [CMake project][0] which does nothing but prints out a few lines of texts for reproducing the crash.

[0]: https://github.com/chchwy/upx-macos13-crash

Here is the commands I used to compressed the executable. Then it crashes after I tried to run the compressed executable.
```sh
% upx -vk ./HelloUPX4 -o./HelloUPX4-compressed
Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2020
UPX 3.96        Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 23rd 2020

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
     82304 ->     12304   14.95%   macho/amd64   HelloUPX4-compressed          

Packed 1 file.

% ./HelloUPX4-compressed 
zsh: segmentation fault  ./HelloUPX4-compressed
%
```

lldb
```sh
% lldb ./HelloUPX4-compressed 
(lldb) target create "./HelloUPX4-compressed"
Current executable set to '/Users/matt/Work/HelloUPX4/build/Debug/HelloUPX4-compressed' (x86_64).
(lldb) run
Process 14499 launched: '/Users/matt/Work/HelloUPX4/build/Debug/HelloUPX4-compressed' (x86_64)
Process 14499 stopped
* thread #1, stop reason = EXC_BAD_ACCESS (code=1, address=0x21090c000)
    frame #0: 0x000000021090c000
error: memory read failed for 0x21090c000
Target 0: (HelloUPX4-compressed) stopped.
(lldb) exit
```

## What should have happened?

It should print the same texts as before it was compressed.

This is the output when it is not compressed.
```
% lldb "HelloUPX4"
(lldb) target create "HelloUPX4"
Current executable set to '/Users/evatu/Work/HelloUPX4/build/Debug/HelloUPX4' (x86_64).
(lldb) run
Process 14516 launched: '/Users/evatu/Work/HelloUPX4/build/Debug/HelloUPX4' (x86_64)
Hello UPX4!
Done.
Process 14516 exited with status = 0 (0x00000000) 
```

## Do you have an idea for a solution?
Actually no, but It was working on macOS 12.

## How can we reproduce the issue?

I made a small CMake project with merely 16 lines code for this bug report. It is configured to generate an amd64 executable.
1. Clone the project at https://github.com/chchwy/upx-macos13-crash and go to the folder
2. `mkdir build`
3. `cd build`
4. `cmake .. -G Xcode`
5. `cmake --build .`
6. `upx -vk ./Debug/HelloUPX4 -o./Debug/HelloUPX4-compressed
7. `./Debug/HelloUPX4-compressed` <= crash here

## Please tell us details about your environment.
<!--- Include as many relevant details about the environment you experienced the bug in -->
* UPX version used (`upx --version`): 3.96
* Host Operating System and version: macOS 13.0 Ventura
* Host CPU architecture: arm64
* Target Operating System and version: same
* Target CPU architecture: amd64 (or called x86_64)

@chchwy On Apple M1 hardware, then MacOS 13.0 (Ventura) with Xcode 14.1 (Nov.1, 2022) is not ready for prime time:
```
## Download from Apple and install  Command_Line_Tools_for_Xcode_14.1.dmg
## Then:
% cmake .. -G Xcode
zsh: command not found: cmake
## Download and install HomeBrew:
% /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
   [[snip]]
% brew --version
Homebrew 3.6.10
Homebrew/homebrew-core (git revision 3e28f99040e; last commit 2022-11-13)
% brew install cmake
% cmake .. -G Xcode
CMake Error:
  Xcode 1.5 not supported.
% cmake --version
cmake version 3.24.3
```
Suggestions?

> @chchwy On Apple M1 hardware, then MacOS 13.0 (Ventura) with Xcode 14.1 (Nov.1, 2022) is not ready for prime time:
> ```
> ## Download from Apple and install  Command_Line_Tools_for_Xcode_14.1.dmg
> ## Then:
> % cmake .. -G Xcode
> zsh: command not found: cmake
> ## Download and install HomeBrew:
> % /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
>    [[snip]]
> % brew --version
> Homebrew 3.6.10
> Homebrew/homebrew-core (git revision 3e28f99040e; last commit 2022-11-13)
> % brew install cmake
> % cmake .. -G Xcode
> CMake Error:
>   Xcode 1.5 not supported.
> % cmake --version
> cmake version 3.24.3
> ```
> 
> Suggestions?

@jreiser 

Run the following command before cmake. Please specify the Xcode.app path to the one you installed.
```
sudo /usr/bin/xcode-select --switch /Applications/Xcode-14.1.0.app
```

Probably you also need to open the Xcode app for the license accepting stuff for the first time.

> Info from #613 : macOS 13 source code has been released https://github.com/apple-oss-distributions/xnu

In addition to my effort using `upx/src/stub/tools/macho-snip/macho-snip.c`, which is somewhat bottom-up (find a minimal runnable executable by starting with /bin/date, then remove `Macho_command`  from it, one-by-one, trying to figure out why E_NOEXEC, E_BADEXEC, etc.) I am also trying a somewhat top-down approach.  Start with the xnu source, compile it, gather all `*.o` into a `*.a` archive library, use that library with a user-mode main program that just does execve(path, args, envp), and use `lldb` to trace execution in user mode of the subroutines from xnu.  The hope is to gather more clues about what is needed, and why, for a packed-by-upx file to execute successfully.

But I encounter problems when I try `https://github.com/apple-oss-distributions/xnu.git`.  Starting with commit `5c2921b07a2480ab43ec66f5b9e41cb872bc554f` of Tue Oct 18 20:56:54 2022 +0000 whose log says xnu-8792.41.9,
I encounter a missing file
```
xnu/libkern/os/log.c:33:10: fatal error: 'os/firehose_buffer_private.h' file not found
#include <os/firehose_buffer_private.h>

xnu/libkern/c++/OSKext.cpp:38:10: fatal error: 'os/firehose_buffer_private.h' file not found
#include <os/firehose_buffer_private.h>
```
Searching the net, I find a plausible substitute  `https://opensource.apple.com/source/libdispatch/libdispatch-1271.100.5/os/firehose_buffer_private.h` which I download as `xnu/libkern/os/firehose_buffer_private.h`.  But `firehose_buffer_private.h` still cannot be found by log.c or OSKext.cpp, even after starting over via `make clean; make`.  Has anybody actually built xnu?

(The xnu source also draws many hundreds of complaints from `-Wunused-but-set-variable`.)

Update Xcode by hand, still get same error.
```
 AppStore > Search Xcode > Update  (3.2 GB)

% xcodebuild -version
Xcode 15.0
Build version 15A240d
% makexcrun: error: sh -c '/Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -find tightbeamc 2> /dev/null' failed with exit code 17664: (null) (errno=No such file or directory)
xcrun: error: unable to find utility "tightbeamc", not a developer tool or in PATH
/Users/jreiser/xnu/makedefs/MakeInc.kernel:286: *** Could not determine xnu version from SDK or KDK! Set RC_DARWIN_KERNEL_VERSION environment variable..  Stop.
make[1]: *** [build_setup_bootstrap_DEVELOPMENT^X86_64^NONE] Error 2
make: *** [all] Error 2
% 
```

--- a/src/p_mach.cpp
+++ b/src/p_mach.cpp
@@ -1491,7 +1491,8 @@ void PackMachBase<T>::unpack(OutputFile *fo)
     fi->readx(rawmseg, mhdri.sizeofcmds);
 
     // FIXME forgot space left for LC_CODE_SIGNATURE;
-    // but canUnpack() sets overlay_offset anyway.
+    if (is_arm64()) overlay_offset += 16384; // LC_CODE_SIGNATURE space
+     but canUnpack() sets overlay_offset anyway.
     //overlay_offset = sizeof(mhdri) + mhdri.sizeofcmds + sizeof(linfo);
 
     fi->seek(overlay_offset, SEEK_SET);

